docs/rfc/rfc-011-custom-indexing.md

```diff
@@ -0,0 +1,216 @@
+disconnect the client. That means that RPC subscribers may lose (miss) events.
+The indexer, however, is "unbuffered", and the publisher will never drop or
+disconnect from it. If the indexer is slow, the publisher will block until it
+returns, to ensure that no events are lost.
+
+In practice, this means that the performance of the indexer has a direct effect
+on the performance of the consensus node: If the indexer is slow or stalls, it
+will slow or halt the progress of consensus. Users have already reported this
+problem even with the built-in indexer (see, for example, [#7247][i7247]).
+Extending this concern to arbitrary user-defined custom indexers gives that
+risk a much larger surface area.
+
+
+## Discussion
+
+Although there is no unique "best" solution to the issues described above,
+there are some specific principles that a solution should include:
+
+1. **A custom indexer should not require integration into Tendermint core.** A
+   developer or node operator can create, build, deploy, and use a custom
+   indexer with a stock build of the Tendermint consensus node.
+
+2. **Custom indexers cannot stall consensus.** An indexer that is slow or
+   stalls cannot slow down or prevent core consensus from making progress.
+
+   The plugin interface must give node operators control over the tolerances
+   for acceptable indexer performance, and the means to detect when indexers
+   are falling outside those tolerances, but indexer failures should "fail
+   safe" with respect to consensus (even if that means the indexer may miss
+   some data, in sufficiently-extreme circumstances).
+
+3. **Custom indexers control which events they index.** A custom indexer is not
+   limited to only the current transaction and block events, but can observe
+   any event published by the node.
+
+4. **Custom indexing is forward-compatible.** Adding new event types or event
+   data to the consensus node should not require existing custom indexers to be
+   rebuilt or modified, unless they want to take advantage of the new data.
+
+5. **Indexers are responsible for answering queries.** An indexer plugin is not
+   required to support the legacy query filter language, nor to be compatible
+   with the legacy RPC endpoints for accessing them.  Any APIs for clients to
+   query a custom index are the responsibility of the indexer, not the node.
+
+### Informal Design Intent
+
+The design principles described above implicate several components of the
+Tendermint node, beyond just the indexer. In the context of [ADR 075][adr075],
+we are re-working the RPC event subscription API to improve some of the UX
+issues discussed above for RPC clients. It is our expectation that a solution
+for pluggable custom indexing will take advantage of some of the same work.
+
+On that basis, the design approach I am considering for custom indexing looks
+something like this (subject to refinement):
+
+1. A custom indexer runs as a separate process from the node.
```

Comment:
I definitely agree with what you've said here: people care more about consensus performance than indexing performance. The way that the current implementation entangles these concerns is indeed massively confusing to users (complicated by the features that hang off of the kv indexer too.) 

I'm a little worried that people kind of want both (who doesn't?), and that if we remove the ability for the indexers to block consensus, then people will *then* be very concerned about indexers (or just subscription processing) being slow/falling behind.

In some ways it might not matter very much: we're saying (I hope I've this right) "instead of having built in indexers, we're going to make event subscription better, and then you can just write your own indexing tools that use good subscriptions." This is is a fine model and I think both lets people build what they want and also allows us to wash our hands of the problem. 

I'm a bit worried that: round tripping a boatload of data through json and adding an extra networking hop might not be super compelling for users (and also might push some users into using us as their message bus (why pipe subscriptions through sidecars and external processes when you could just consume them directly"?)

Reply:
> I'm a little worried that people kind of want both (who doesn't?), and that if we remove the ability for the indexers to block consensus, then people will _then_ be very concerned about indexers (or just subscription processing) being slow/falling behind.

We already have a lot of folks concerned about subscription processing falling behind, so I think we can actually turn this around: _If_ we can make subscription processing acceptable for users, then it should also (I claim) suffice for indexing.

If we can't, then we might make a different choice here—but then we should probably also do the same thing for other types of subscription as well.

> In some ways it might not matter very much: we're saying (I hope I've this right) "instead of having built in indexers, we're going to make event subscription better, and then you can just write your own indexing tools that use good subscriptions." 

Yes, that's more or less the summary of the "intended implementation" I described below. Part of my point in raising this in the RFC is specifically to shake out these kinds of concerns.

> I'm a bit worried that: round tripping a boatload of data through json and adding an extra networking hop might not be super compelling for users (and also might push some users into using us as their message bus 

Well, the "networking" hop is on the loopback unless one is doing something weird, so that cost is essentially no higher than writing to a file (which is what any in-process indexer would be doing anyway). 

Conversion to JSON is a cost, but not much of one: We're already doing that conversion for event subscribers anyway, so doing it for the indexer doesn't add any work for the node itself. An indexer would probably use a different format—but then it's already stuck with a conversion issue. 

Consider the PSQL sink for example: Some of the data it stores are just protobuf messages, but most of it is picked out and converted into strings to poke into SQL. The cost difference is almost certainly negligible compared to the overhead of doing the write.  Anecdotally, a lot of people writing tools in this ecosystem seem to dislike the need to deal with protobufs anyway, though possibly that segment doesn't overlap as strongly for folks wanting to write their own indexers.

> (why pipe subscriptions through sidecars and external processes when you could just consume them directly"?)

I tried to address the "consume them directly" concern: That requires living in-process with the node, and then we have to have some kind of fancy dynamic plugin architecture at build time. I'm fairly convinced that's a bad idea, but if there's a better argument for it I'd be interested to address it.

Reply:
Some of our production services rely on the tx indexing events, if there are some tx events missing, it will reflect issues to our production services, e.g. missing tx on the blockchain dashboard built on top of the Evmos, etc. I am worried about it. It will cause the system operating overhead.

Reply:
I agreed with you. The system cannot have both things perfect.
If we accept the system has a small chance of missing the events. Would it be possible to send a request to replay the missing events? How do we get the missing events during the run-time?

Reply:
> I agreed with you. The system cannot have both things perfect. If we accept the system has a small chance of missing the events. Would it be possible to send a request to replay the missing events? How do we get the missing events during the run-time?

There are two sides to missing events:

1. An event may be lost because the indexer did not read it before it expired from the node. For example, suppose the node buffers the last 2 minutes of events, but the indexer is too slow and an event from 3 minutes ago was dropped before the indexer saw it.

   In this case, the event is truly "dropped", but that will only happen if the indexer is truly incredibly slow. That could happen, but the node operator would have some control over how likely it is (a resource/reliability tradeoff, memory for replay window).

2. An event may be lost because the indexer crashed and restarted, or lost its connection to the node. In this case, the indexer would be able to reconnect and replay missed events as long as they're not too old (where "too old" is a configuration setting, as noted above).

So in situation (2), an indexer could replay, but in situation (1) it could not.

Today, the only way we can miss a node is situation (1), if the node crashes before some events are captured by the indexer. The down side, however, is that to enforce that guarantee requires the node to block until the indexer is done recording the event. That's why slow indexing can stall or slow consensus.

Reply:
So in case 1), the node-op requires to know how to set it properly. For the indexer data storage, it might work well with the original settings at the beginning, but slow down when the storing data scale-up later. (The KV database might execute heavy compaction occasionally when data scale-up, it could stale the disk a while) When it happens, the settings might need to be re-tuned. And events will be missing at that time I think.

in case 2). The indexer should have the ability to recover the service itself to avoid the long disconnect from the node. right? Otherwise, a few mins buffer might not enough for the node-op to notice and to recover it (if the indexer crashed).

I still think it's good to have a way to catch up the missing events (already dropped by the node somehow). At least to leave a capability do it later.

Reply:
> So in situation 1), the node-op requires to know how to set it properly. 

Yes, that's right. Assuming we use a time-window type setting, the operator can estimate a reasonable value based on knowledge of the block profile: For example, if the block time is x minutes and blocks are typically around n transactions, it is fairly easy to roughly predict the number of event items: Each transaction generates a transaction event, and each block generates an event. (There may be multiple ABCI metadata per item, but the indexer gets each item as a unit).

Combined with some simple benchmarks of the indexer, this approach would make it fairly easy for operators to estimate a reasonable window and safety margin.

> For the indexer data storage, it might work well with the original settings at the beginning, but slow down when the storing data scale-up later. 

That's true. That's arguably a problem to fix in the indexer, not the node: Right now the built-in KV indexer has this issue, which I think is partly why folks want to write their own custom indexers. 

That said, I wouldn't generalize too much from the KV indexer. Its implementation is fairly ad hoc, and I don't think it was ever really intended for serious production use.

> in situation 2). The indexer should have the ability to recover the service itself to avoid the long disconnect from the node. right? 

Yes. The idea is to run the indexer "close" to the node—probably on the same machine or in a nearby container—just as you would for, say, an auth sidecar, a sentry proxy, or any other essential service. The expected mode of failure is a process crash, not arbitrary network delay. 

The usual ops tactics should suffice: Alert on process crash, alert on weird load metrics, etc.

> Otherwise, a few mins buffer might not enough for the node-op to notice it and to recover it (if the indexer crashed).

An operator can set as much buffer as they find reasonable. If they want to provision the memory for it, they could set it to 30 minutes if they want. It's a tradeoff, of course; you spend memory to get resilience, but there's always an upper bound.

> I still think it's good to have a way to catch up on the missing events (already dropped by the node somehow). At least leaves a capability to implement it later.

This touches on a key point that I should clarify in the doc: The node does not persist events separately from the indexer. The indexer _is_ the component responsible for persisting events. In nodes that don't run an indexer, there is no way to "replay" events. The node stores the blocks, and of course one could write a tool to go back through those blocks, but the indexer is responsible for the app metadata.

So what you're describing here is not available: If the node already had a log of all the events, we wouldn't need an indexer at all.

Reply:
I added some text to clarify that the node does not index application metadata separately from the indexer itself.